pkg/metrics/server.go

```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
 		Handler: router,
 	}
 
-	go func() {
-		if err := srv.ListenAndServeTLS(tlsCRT, tlsKey); err != nil {
-			klog.Errorf("error starting metrics server: %v", err)
+	return srv
+}
+
+func startServer(srv *http.Server, stopCh <-chan struct{}) {
+	klog.Infof("starting metrics server")
+	if err := srv.ListenAndServeTLS(tlsCRT, tlsKey); err != nil && err != http.ErrServerClosed {
+		klog.Errorf("error from metrics server: %v", err)
+	}
+}
+
+func stopServer(srv *http.Server) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := srv.Shutdown(ctx); err != nil {
+		klog.Errorf("error or timeout stopping metrics listener: %v", err)
+	} else {
+		// TODO delete debugging lines
+		klog.Infof("stopping metrics server without error. Err: %v", err)
+	}
+}
+
+// RunServer starts the server, and watches the tlsCRT and tlsKey for certificate changes
+func RunServer(port int, stopCh <-chan struct{}) {
+	srv := buildServer(port)
+	if srv == nil {
+		return
+	}
+
+	go startServer(srv, stopCh)
+
+	// Set up and start the file watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Errorf("failed to create watcher: %v", err)
+	}
+	defer watcher.Close()
+	if err := watcher.Add(certsDir); err != nil {
+		klog.Errorf("failed to add %v to watcher: %v", certsDir, err)
+	} else {
+		klog.Infof("watching: %v", certsDir)
+	}
+
+	origCertChecksum := checksumFile(tlsCRT)
+	origKeyChecksum := checksumFile(tlsKey)
+
+	for {
+		select {
+		case <-stopCh:
+			stopServer(srv)
+		case event, ok := <-watcher.Events:
+			if !ok {
+				klog.Errorf("file watcher events channel closed")
+			}
+
+			klog.Infof("file: %v, event: %v", event.Name, event.Op)
```

Comment:
Good catch. I made this V(2) as it may be useful for future debugging. I also made it so that we will skip the checksum step if the file event is chmod or remove. The other event types are create, write, and rename.